### PR TITLE
Add TricrystalTripleJunctionIC to phase_field module

### LINKED ISSUE #7196: Add TricrystalTripleJunctionIC initial condition to phase-field module
### Description of the enhancement or error report
This ic allows the user to specify two angles in degrees to simulate three grains meeting at a triple junction.

### Rationale for the enhancement or information for reproducing the error
The ability to simulate a single triple junction is helpful to analyze directly what is happening at a triple junction.  Also takes a step forward to testing the model implemented in MOOSE for anisotropic grain growth.

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

### PR DESCRIPTION
**Pull request by jarinfrench**
### Complete each of the following items before creating a PR
- Include any relevant design information for your change
- Follow [Coding Standards](http://mooseframework.org/wiki/CodeStandards/)
- Submit one or more [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/) or modify existing test cases
  - TricrystalTripleJunctionIC.i passes the test TricrystalTripleJunctionIC in the initial_conditions tests file
- Reference or close a specific issue (refs # or closes #)
  - Closes #7196

### CODE REVIEW

**Comment by permcody** [comment]
Oops - you "pulled" instead of "rebased".  Please rebase your work on devel and push again.

**Comment by jarinfrench** [comment]
Any ideas why the modules debug test caused my test to crash?

I'm also struggling to determine what's causing the intel test to fail.  It feels like I've missed something in learning how to test an initial condition.  Any thoughts?

**Comment by permcody** [comment]
Well something is just plain wrong but it's not immediately obvious to me what it is. The debug failure tells us that we are doing something wrong with iterators somewhere. We turn on "safe iterators" in debug mode when we are compiling with GCC on Linux and it checks everything! Something wrong can me iterating past the end, or it can mean that we've mixed up iterators and are comparing the wrong one to the end or something. Unfortunately we don't have a lot of good information as to where this is happening. It is in the FeatureFloodCount object so it's probably not your fault.

As to the Intel failure, I believe this is independent of the other problem. It might just be differences in roundoff between Intel and GCC compilers but it seems a little bigger than I would expect. I'll take a look at the debug iterator problem as I suspect that's me. You should probably wait on the Intel just in case it's related.

File: modules/phase_field/include/ics/TricrystalTripleJunctionIC.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+#include <string>
+
+// Forward Declarations
+class TricrystalTripleJunctionIC;
+
+template<>
+InputParameters validParams<TricrystalTripleJunctionIC>();
+
+/**
+* TricrystalTripleJunctionIC creates a 3-grain structure with a triple junction in the center
```
Comment (dschwen) [inline modules/phase_field/include/ics/TricrystalTripleJunctionIC.h]: indent

File: modules/phase_field/include/ics/TricrystalTripleJunctionIC.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+*/
+class TricrystalTripleJunctionIC : public InitialCondition
+{
+public:
+  /**
+  * Constructor
+  *
+  * @param parameters The parameters object holding data for the class to use.
+  */
```
Comment (dschwen) [inline modules/phase_field/include/ics/TricrystalTripleJunctionIC.h]: please remove boilerplate comment

**Comment by jarinfrench** [comment]
@dschwen I'm not quite sure what you want me to indent.  Are you wanting me to in general check the indentation of the code, or look specifically at those lines you highlighted?

File: modules/phase_field/include/ics/TricrystalTripleJunctionIC.h (hunk 1 above)
Comment (dschwen) [inline modules/phase_field/include/ics/TricrystalTripleJunctionIC.h]: Should be 
```
/**
 * Blah
 */
```

not

```
/**
* Blah
*/
```

File: modules/phase_field/tests/initial_conditions/TricrystalTripleJunctionIC.i [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    GBmob0 = 2.5e-6 #m^4(Js) for copper from Schoenfelder1997
+    Q = 0.23 #eV for copper from Schoenfelder1997
+    GBenergy = 0.708 #J/m^2 from Schoenfelder1997
+  [../]
+[]
+
+[Postprocessors]
+  # Scalar postprocessors
+  [./grain_tracker]
+    type = GrainTracker
```
Comment (dschwen) [inline modules/phase_field/tests/initial_conditions/TricrystalTripleJunctionIC.i]: Could you try using `FauxGrainTracker` here instead. Maybe we can cirumvent the iterator issue in FFC for now and get this merged.

Comment (jarinfrench) [inline modules/phase_field/tests/initial_conditions/TricrystalTripleJunctionIC.i]: Oops, didn't see this before I pushed the changes from above.  Yes, I will modify the input file to work with ```FauxGrainTracker``` instead.  I'm still learning how to use ```GrainTracker```, so that is why there seems to be a lot of fluff.  I'll fix that as well.

**Comment by dschwen** [comment]
Well, now it exodiffs

**Comment by jarinfrench** [comment]
I don't understand why it fails the Exodiff test.  Is the Exodiff due to how I created the initial condition?

File: modules/phase_field/include/ics/TricrystalTripleJunctionIC.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  unsigned int _op_num;
+  unsigned int _op_index;
+
+  Point _bottom_left;
+  Point _top_right;
+  Point _range;
+  Real _theta1; // Angle of first grain at triple junction in radians
+  Real _theta2; // Angle of third grain at triple junction in radians
+  Real _tan_theta1; //tangent of the first angle after a shift of pi/2
+  Real _tan_theta2; //tangent of the second angle after a shift of pi/2
```
Comment (dschwen) [inline modules/phase_field/include/ics/TricrystalTripleJunctionIC.h]: Please use Doxygen style comments for all of these (i.e. triple slash `///` and on the line preceding the member declaration.

File: modules/phase_field/src/ics/TricrystalTripleJunctionIC.C [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+  // Change the third angle to be measured from the +x axis
+  _theta2 = _theta2 - libMesh::pi / 2.0;
+
+  // Set up domain bounds with mesh tools
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+  {
+    _bottom_left(i) = _mesh.getMinInDimension(i);
+    _top_right(i) = _mesh.getMaxInDimension(i);
+  }
+  _range = _top_right - _bottom_left;
```
Comment (dschwen) [inline modules/phase_field/src/ics/TricrystalTripleJunctionIC.C]: `_bottom_left` and `_top_right` should not be member variables. They are never used again. `_range` should be `_center` or `_junction`, and it should be a parameter the user can specify.

File: modules/phase_field/src/ics/TricrystalTripleJunctionIC.C [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+TricrystalTripleJunctionIC::value(const Point & p)
+{
+  /*
+   * This does all the work to create a triple junction that looks like the letter Y
+   */
+  Real dist_left; // The distance from the point to the line specified by _theta1
+  Real dist_right; // The distance from the point to the line specified by _theta2
+
+  // Check if the point is above or below the left-most line
+  // Function to use is y = max_y/2.0 + (x - max_x/2.0) * std::tan(libMesh::pi/2.0 - _theta1)
```
Comment (dschwen) [inline modules/phase_field/src/ics/TricrystalTripleJunctionIC.C]: This comment is wrong. The calculation is _not_ using a `max_x` or `max_y` value, it is using something like `max-min`.

File: modules/phase_field/src/ics/TricrystalTripleJunctionIC.C [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+  }
+
+  else // all other cases
+  {
+    if ((dist_left <= 0 && dist_center <= 0 && _op_index == 1) || // First grain
+        (dist_right <= 0 && dist_center > 0 && _op_index == 2) || // Second grain
+        (((dist_left > 0 && dist_center < 0) || (dist_right > 0 && dist_center >= 0)) && _op_index == 3)) // Third grain
+       return 1.0;
+    else
+       return 0.0;
```
Comment (dschwen) [inline modules/phase_field/src/ics/TricrystalTripleJunctionIC.C]: It would not be particularly hard to add the capability for smooth interfaces, but maybe we leave that for another PR.

File: modules/phase_field/tests/initial_conditions/TricrystalTripleJunctionIC.i [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+  [../]
+[]
+
+[Executioner]
+  type = Transient
+  scheme = bdf2
+  solve_type = 'PJFNK'
+
+  petsc_options_iname = '-pc_type -pc_hypre_type -ksp_gmres_restart -mat_mffd_type'
+  petsc_options_value = 'hypre boomeramg 31 ds'
```
Comment (dschwen) [inline modules/phase_field/tests/initial_conditions/TricrystalTripleJunctionIC.i]: I bet the problem solves just as well without all those PETSc options.

File: modules/phase_field/tests/initial_conditions/TricrystalTripleJunctionIC.i [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+  num_steps = 1
+[]
+
+[Outputs]
+  exodus = true # Outputs to the Exodus file format
+  execute_on = 'final'
+[]
+
+[Problem]
+  solve = false
```
Comment (dschwen) [inline modules/phase_field/tests/initial_conditions/TricrystalTripleJunctionIC.i]: Haha, yeah, simplify the Executioner block down to 

```
[Executioner]
  type = Transient
[]
```

there is no solve. Our test files should be as simple as possible.

File: modules/phase_field/src/ics/TricrystalTripleJunctionIC.C (hunk 6 above)
Comment (jarinfrench) [inline modules/phase_field/src/ics/TricrystalTripleJunctionIC.C]: Demonstrating my ignorance a bit here, does the mesh not start at ```(0,0,0)``` and end at some ```max_value```?

File: modules/phase_field/src/ics/TricrystalTripleJunctionIC.C (hunk 5 above)
Comment (jarinfrench) [inline modules/phase_field/src/ics/TricrystalTripleJunctionIC.C]: In trying to use a member variable that is user-defined, I run into the issue where I don't know how to set the default value.  If I'm going to have it be user-defined, I was thinking of having the default be the center of the mesh, but trying to use the ```_mesh``` variable doesn't work as a default.  Is there a way to go about doing that?  Or should I just use ```Point(0, 0, 0)``` as my default?

Comment (dschwen) [inline modules/phase_field/src/ics/TricrystalTripleJunctionIC.C]: The variable would not even need to have a default (it could use your mesh center calculation if the user does not supply it). Otherwise I'd go with `Point()` (or `Point(0.0, 0.0, 0.0)` for a more verbose version)

**Comment by dschwen** [comment]
@jarinfrench the intel failure may be mitigated by slightly resizing your mesh! Use an **odd** number of elements in each direction. Your IC is set up to sit in the _center_ of the cell, that means you have value changes _exactly_ on the nodes on the center line. If you use an odd number of cells the center line is between nodes and there is _no_ ambiguity as to where a node belongs to the left or right grain.

**Comment by jarinfrench** [comment]
That seems to address part of the problem, but somehow it still exodiffs.  Do I need to make the mesh more course?  Or would making the mesh finer work better?  I'm going to try with a courser mesh and see how that plays out.

**Comment by dschwen** [comment]
Figure out where Node 79 is located in your mesh.

**Comment by jarinfrench** [comment]
How do I do that?

**Comment by dschwen** [comment]
You could use paraview with your output.

**Comment by dschwen** [comment]
Try changing the angles a bit. One of the upper boundaries may still hit a node exactly.

File: modules/phase_field/src/ics/TricrystalTripleJunctionIC.C [hunk 10]
```diff
@@ -37,6 +38,23 @@ TricrystalTripleJunctionIC::TricrystalTripleJunctionIC(const InputParameters & p
...
 
+  // Make sure that _junction is in the domain
+  for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+  {
+    if ((_mesh.getMinInDimension(i) > _junction(i)) || (_mesh.getMaxInDimension(i) < _junction(i)))
+      mooseError("Triple junction out of bounds");
+  }
+
+  // Default junction point is the center
+  if (!parameters.isParamSetByUser("junction"))
```
Comment (dschwen) [inline modules/phase_field/src/ics/TricrystalTripleJunctionIC.C]: `isParamValid` would be the correct call here (`isParamSetByUser` makes sense if a default was supplied - then you can check if it was overridden by the user)

File: modules/phase_field/src/ics/TricrystalTripleJunctionIC.C [hunk 11]
```diff
@@ -37,6 +38,23 @@ TricrystalTripleJunctionIC::TricrystalTripleJunctionIC(const InputParameters & p
...
+  {
+    for (unsigned int i = 0; i < LIBMESH_DIM; ++i)
+      _junction(i) = _mesh.getMaxInDimension(i) / 2.0;
```
Comment (dschwen) [inline modules/phase_field/src/ics/TricrystalTripleJunctionIC.C]: Wait, that assumes that your mesh starts at (0,0,0)...

**Comment by jarinfrench** [comment]
Thanks, I think I've found that Node 79 (is this the global node ID?) is located at ```(546,546,0)```, which is on a boundary between grains.  I'm just going to change the angle of that boundary and see what comes of it.
